compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -10774,40 +10774,80 @@ private STNode parseTupleMemberTypeDescList() {
     }
 
     private STNode parseTupleTypeMembers(STNode typeDesc, List<STNode> typeDescList) {
-        STToken nextToken;
+        STNode tupleMemberRhs = parseTypeDescInTupleRhs();
+        if (tupleMemberRhs != null) {
+            typeDesc = STNodeFactory.createRestDescriptorNode(typeDesc, tupleMemberRhs);
+        }
+
         // Parse the remaining type descs
-        nextToken = peek();
-        STNode tupleMemberRhs;
-        while (!isEndOfTypeList(nextToken.kind)) {
+        while (!isEndOfTypeList(peek().kind)) {
             tupleMemberRhs = parseTupleMemberRhs();
             if (tupleMemberRhs == null) {
                 break;
             }
-            if (tupleMemberRhs.kind == SyntaxKind.ELLIPSIS_TOKEN) {
-                typeDesc = STNodeFactory.createRestDescriptorNode(typeDesc, tupleMemberRhs);
+
+            if (typeDesc.kind == SyntaxKind.REST_TYPE) {
+                typeDesc = invalidateTypeDescAfterRestDesc(typeDesc);
                 break;
             }
+
             typeDescList.add(typeDesc);
             typeDescList.add(tupleMemberRhs);
-            typeDesc = parseTypeDescriptor(ParserRuleContext.TYPE_DESC_IN_TUPLE);
-            nextToken = peek();
+            typeDesc = parseMemberDescriptor();
         }
 
         typeDescList.add(typeDesc);
         return STNodeFactory.createNodeList(typeDescList);
     }
 
+    private STNode parseMemberDescriptor() {
+        STNode typeDesc = parseTypeDescriptor(ParserRuleContext.TYPE_DESC_IN_TUPLE);
+        STNode tupleMemberRhs = parseTypeDescInTupleRhs();
+
+        if (tupleMemberRhs != null) {
+            return STNodeFactory.createRestDescriptorNode(typeDesc, tupleMemberRhs);
+        }
+
+        return typeDesc;
+    }
+
+    private STNode invalidateTypeDescAfterRestDesc(STNode restDescriptor) {
+        while (!isEndOfTypeList(peek().kind)) {
+            STNode tupleMemberRhs = parseTupleMemberRhs();
+            if (tupleMemberRhs == null) {
+                break;
+            }
+
+            restDescriptor = SyntaxErrors.cloneWithTrailingInvalidNodeMinutiae(restDescriptor, tupleMemberRhs, null);
+            restDescriptor = SyntaxErrors.cloneWithTrailingInvalidNodeMinutiae(restDescriptor, parseMemberDescriptor(),
+                    DiagnosticErrorCode.ERROR_TYPEDESC_AFTER_REST_DESCRIPTOR);
+        }
+
+        return restDescriptor;
+    }
+
     private STNode parseTupleMemberRhs() {
         switch (peek().kind) {
             case COMMA_TOKEN:
                 return parseComma();
+            case CLOSE_BRACKET_TOKEN:
+                return null;
+            default:
+                recover(peek(), ParserRuleContext.TUPLE_TYPE_MEMBER_RHS);
+                return parseTupleMemberRhs();
+        }
+    }
+
+    private STNode parseTypeDescInTupleRhs() {
+        switch (peek().kind) {
```

Comment:
Shall we extract `peek()` to a separate variable and reuse it at the recovery call?